### PR TITLE
[`Ernie 4.5`] Add ernie text models

### PR DESCRIPTION
**Pull request by vasqu**
Adding the Ernie 4.5 suite of models. 

Progress:
- [x] Ernie 4.5 pure text model (0.3B)
- [x] MoE Ernie
  - [x] Loading check with untied weights (tested on a dummy model)
  - [x] TP tests 
    - [x] Failing with tied weights, needs to be fixed then it's done 
  - [x] Correction bias clarification
    - Following the paddle code instead of remote - added a note so subject to change  
  - [x] Update configs on hub
  - [ ] (MTP support in training)
- [x] Integration test ^ (needs slow runs to cross check)
- [x] Check whether the MoE also need a rotation conversion (0.3B modeling files differ to the other ones regarding RoPE)
  - Yes, they do - turns out they use a similar trick as I did in 393c2c772aebeac0c6816bd2086c600254161ea5
  - Adapted from GLM as they do the same RoPE style as well
- [x] Fixup tokenization
  - [x] ~[Conversion](https://gist.github.com/vasqu/7828357fd24929f4d7a51202b0801c3e)~ see `convert...tokenizer`
  - [x] Update on the hub
- [x] Docs (might need updates based on the tokenizer ^)
- [x] Update original hub on baidu side --> tokenizer + configs

New/Followup PR:
- [ ] MoE Ernie VL
  - [ ] MoE is different (not allowing for the original MoE formula (Mixtral-based)?)
    - [ ] It can have different capacities
    - [ ] Different gating :eyes: 
  - [ ] 3D RoPE in image and text (with different RoPE formulation ~GLM style, even/odd instead of half/half)
  - [ ] Miscellaneous, as in the other remote code
    - [ ] Attention 
    - [ ] RMS norm
    - [ ] Residual
    - [ ] Proper padding support etc.

### CODE REVIEW

**Comment by vasqu** [comment]
run-slow: ernie4_5

**Comment by ArthurZucker** [comment]
Let's go! 🚀

File: docs/source/en/model_doc/ernie4_5.md [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+
+# Ernie 4.5
+
+## Overview
+
+The Ernie 4.5 model was released in the [Ernie 4.5 Model Family](https://ernie.baidu.com/blog/posts/ernie4.5/) release by baidu.
+This family of models contains multiple different architectures and model sizes. This model in specific targets the base text
+model without mixture of experts (moe) with 0.3B parameters in total. It uses the standard [Llama](./llama.md) at its core.
+
+Other models from the family can be found at [Ernie 4.5 MoE](./elrnie4_5_moe.md).
```
Comment (ZeyuChen) [inline docs/source/en/model_doc/ernie4_5.md]: ./elrnie4_5_moe.md -> ./ernie4_5_moe.md

File: docs/source/en/model_doc/ernie4_5_moe.md [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+
+## Overview
+
+The Ernie 4.5 MoE model was released in the [Ernie 4.5 Model Family](https://ernie.baidu.com/blog/posts/ernie4.5/) release by baidu.
+This family of models contains multiple different architectures and model sizes. This model in specific targets the base text
+model with mixture of experts (moe) - one with 21B total, 3B active parameters and another one with 300B total, 47B active parameters.
+It uses the standard [Llama](./llama.md) at its core combined with a specialized MoE based on [Mixtral](./mixtral.md) with additional shared
+experts.
+
+Other models from the family can be found at [Ernie 4.5](./elrnie4_5.md).
```
Comment (ZeyuChen) [inline docs/source/en/model_doc/ernie4_5_moe.md]: ./elrnie4_5.md-> ernie4_5.md

**Review by ZeyuChen — COMMENTED** [review]
I found some document link errors.

File: docs/source/en/model_doc/ernie4_5.md (hunk 1 above)
Comment (vasqu) [reply docs/source/en/model_doc/ernie4_5.md]: Wow, that's a bit embarrassing :sweat_smile: thanks for catching!

**Comment by Meiyim** [comment]
`MoE Ernie VL` use `modality-isolation` routing (which is different from mainstream approaches)

should determin token-type in advance. (e.g. use `token_type_ids = input_ids >100295`)

then apply standart moe in two type of token：

```python
text_hidden =  hidden[token_type_ids==0]
text_router_logits = self.text_gate(text_hidden)
text_hidden = MoE(text_router_logits, text_hidden, self.text_experts)


visual_hidden =  hidden[token_type_ids==0]
visual_router_logits = self.text_gate(text_hidden)
visual_hidden = MoE(visual_router_logits, visual_hidden, self.visual_experts)


layer_output_hidden = torch.empty_like(hidden)
layer_output_hidden[token_type_ids==0] =  text_hidden
layer_output_hidden[token_type_ids==1] =  visual_hidden
```

**Comment by vasqu** [comment]
Thanks, that clarifies a few things :)

File: src/transformers/models/ernie4_5/tokenization_ernie4_5.py
Comment (ArthurZucker) [inline src/transformers/models/ernie4_5/tokenization_ernie4_5.py:1]: can be removed and use llama

Comment (ArthurZucker) [reply src/transformers/models/ernie4_5/tokenization_ernie4_5.py:1]: as we have support for abitrary special tokens

File: src/transformers/models/ernie4_5/tokenization_ernie4_5_fast.py
Comment (ArthurZucker) [inline src/transformers/models/ernie4_5/tokenization_ernie4_5_fast.py:1]: same here I think we added support for any additional tokens

File: src/transformers/models/ernie4_5/modular_ernie4_5.py [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+            The 'unsqueeze_dim' argument specifies the dimension along which to unsqueeze cos[position_ids] and
+            sin[position_ids] so that they can be properly broadcasted to the dimensions of q and k. For example, note
+            that cos[position_ids] and sin[position_ids] have the shape [batch_size, seq_len, head_dim]. Then, if q and
+            k have the shape [batch_size, heads, seq_len, head_dim], then setting unsqueeze_dim=1 makes
+            cos[position_ids] and sin[position_ids] broadcastable to the shapes of q and k. Similarly, if q and k have
+            the shape [batch_size, seq_len, heads, head_dim], then set unsqueeze_dim=2.
+    Returns:
+        `tuple(torch.Tensor)` comprising of the query and key tensors rotated using the Rotary Position Embedding.
+    """
+    # glm rope style (with full dim) and full precision
```
Comment (ArthurZucker) [inline src/transformers/models/ernie4_5/modular_ernie4_5.py:69]: yep!

File: src/transformers/models/ernie4_5/modular_ernie4_5.py [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+
+    return q_embed.to(original_dtype), k_embed.to(original_dtype)
+
+
+class Ernie4_5MLP(LlamaMLP):
+    def __init__(self, config: Ernie4_5Config):
+        super().__init__()
+        del self.gate_proj
+        del self.up_proj
+        del self.down_proj
```
Comment (ArthurZucker) [inline src/transformers/models/ernie4_5/modular_ernie4_5.py]: tell us if it does not work, in general the `x = y` should allow this

File: src/transformers/models/ernie4_5/modular_ernie4_5.py [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+        self.down_proj = nn.Linear(self.intermediate_size, self.hidden_size, bias=config.use_bias)
+
+
+class Ernie4_5Attention(LlamaAttention):
+    def __init__(self, config: Ernie4_5Config, layer_idx: int):
+        super().__init__(config, layer_idx)
+        del self.q_proj
+        del self.k_proj
+        del self.v_proj
+        del self.o_proj
```
Comment (ArthurZucker) [inline src/transformers/models/ernie4_5/modular_ernie4_5.py]: same

**Review by ArthurZucker — COMMENTED** [review]
A small review but already looking good!

File: src/transformers/models/ernie4_5/tokenization_ernie4_5.py
Comment (vasqu) [reply src/transformers/models/ernie4_5/tokenization_ernie4_5.py:1]: Changed to adding them afterwards, I think you meant the `system...end` tokens for example. However, after a deep dive, I think there are two options to align with how the remote tokenizer behaves:
1. Decoding of the slow tokenizer is wrongly done; see the 2 methods `_decode` and `convert_tokens_to_string`. With the fast tokenizer, I had no issues matching the original decoding while the slow tokenizer resulted in wrong decoding. 
2. Apply the changes I did to the slow tokenizer globally.
  - I am not entirely sure what it would break.
  - I thought it's safer to treat it as a local special case instead of globally.

I thought about maybe switching legacy added with new adding (which I haven't found a way to do so yet), but after a while I noticed that would not solve all issues either (i.e. `_decode` would still need to be overwritten).

Side note, for reproducability:
```python
from transformers import AutoTokenizer, Ernie4_5Tokenizer, Ernie4_5TokenizerFast


model_name = "baidu/ERNIE-4.5-0.3B-Base-PT"

# the current version of the remote code ignores any special token, so it needs to be modified
# (this behavior was only introduced recently)
rem_tok = AutoTokenizer.from_pretrained(model_name, trust_remote_code=True)
hf_tok = Ernie4_5Tokenizer.from_pretrained("AntonV/Ernie4_5Tokenizer")
hf_fast_tok = Ernie4_5TokenizerFast.from_pretrained("AntonV/Ernie4_5Tokenizer")

sequences = [
    "Transformers (formerly known as pytorch-transformers and pytorch-pretrained-bert) provides "
    "general-purpose architectures (BERT, GPT-2, RoBERTa, XLM, DistilBert, XLNet...) for Natural "
    "Language Understanding (NLU) and Natural Language Generation (NLG) with over 32+ pretrained "
    "models in 100+ languages and deep interoperability between Jax, PyTorch and TensorFlow.",
    " BERT is designed to pre-train deep bidirectional representations from unlabeled text by jointly "
    "conditioning on both left and right context in all layers.",
    "▁The quick brown fox jumps over the lazy dog.",
    "The following string should be properly encoded: Hello.",
    "Hey<eos>. \t\t \n\nyou  é  @#😈  🤗!       , 1234 15 5,61",
    "This is a test",
    "Hi  Hello",
    "Hi   Hello",
    "",
    " ",
    "  ",
    " Hello",
    ".",
    "ا .",
    " <s> Hello<s> how",
    "<|AUDIO_PLACEHOLDER|>test ok<|AUDIO_PLACEHOLDER|>",
    "<|AUDIO_PLACEHOLDER|>test <|AUDIO_PLACEHOLDER|>ok",
    "<|AUDIO_PLACEHOLDER|>test <|AUDIO_PLACEHOLDER|> ok",
    "test<|AUDIO_PLACEHOLDER|> <|AUDIO_PLACEHOLDER|> ok",
    "test <|AUDIO_PLACEHOLDER|> <|AUDIO_PLACEHOLDER|> ok",
]

rem_model_inputs = rem_tok(sequences)
hf_model_inputs_1 = hf_tok(sequences)
hf_model_inputs_2 = hf_fast_tok(sequences)

# quick and dirty checks: 1. legacy is needed for encoding 2. decoding needs the overwrites
for input_ids_rem, input_ids_hf, input_ids_hf_fast in zip(rem_model_inputs.input_ids, hf_model_inputs_1.input_ids, hf_model_inputs_2.input_ids):
    print(input_ids_rem)
    print(hf_fast_tok.decode(input_ids_hf_fast))
    print(hf_tok.decode(input_ids_hf))
    print(rem_tok.decode(input_ids_rem))
    print(input_ids_hf_fast == input_ids_hf)
    print(input_ids_rem == input_ids_hf_fast)
    print(input_ids_rem == input_ids_hf)
    print()
```

File: src/transformers/models/ernie4_5/tokenization_ernie4_5_fast.py
Comment (vasqu) [reply src/transformers/models/ernie4_5/tokenization_ernie4_5_fast.py:1]: See slow tokenizer reasons, having two different classes would not be the most stellar experience then imo. Especially for loading / conversions.

File: src/transformers/models/ernie4_5/modular_ernie4_5.py (hunk 4 above)
Comment (vasqu) [reply src/transformers/models/ernie4_5/modular_ernie4_5.py]: Nope, works and not sure why I thought it wouldnt tbh :thinking:

File: src/transformers/models/ernie4_5/tokenization_ernie4_5.py
Comment (vasqu) [reply src/transformers/models/ernie4_5/tokenization_ernie4_5.py:1]: Discussed internally, giving priority to fast tokenizers as they work as expected and disabling slow.

File: src/transformers/models/ernie4_5/convert_ernie4_5_tokenizer.py [hunk 6]
```diff
@@ -0,0 +1,58 @@
+import argparse
```
Comment (ArthurZucker) [inline src/transformers/models/ernie4_5/convert_ernie4_5_tokenizer.py:15]: missing licence!

File: src/transformers/models/ernie4_5_moe/modular_ernie4_5_moe.py [hunk 7]
```diff
@@ -0,0 +1,333 @@
...
+        self.top_k = config.moe_k
+
+        # correction bias (yes it seems to be a typo with statics <> statistics)
+        self.moe_statics = Ernie4_5_MoEStatics(config)
+
+        # gating
+        self.gate = nn.Linear(config.hidden_size, config.moe_num_experts, bias=False, dtype=torch.float32)
+        self.experts = nn.ModuleList(
+            [Ernie4_5_MoEMLP(config, config.moe_intermediate_size) for _ in range(config.moe_num_experts)]
+        )
```
Comment (ArthurZucker) [inline src/transformers/models/ernie4_5_moe/modular_ernie4_5_moe.py:125]: lets avoid module lists we should go with llama4 like moe!

File: src/transformers/models/ernie4_5_moe/modular_ernie4_5_moe.py [hunk 8]
```diff
@@ -0,0 +1,333 @@
...
+        self.norm_min = config.moe_norm_min
+
+        # (optional) shared experts for all forwards
+        self.shared_experts = None
```
Comment (ArthurZucker) [inline src/transformers/models/ernie4_5_moe/modular_ernie4_5_moe.py:129]: any models ahve this?

File: src/transformers/models/ernie4_5_moe/modular_ernie4_5_moe.py [hunk 9]
```diff
@@ -0,0 +1,333 @@
...
+            # states by `routing_weights` on the corresponding tokens (top-1 and top-2)
+            current_state = hidden_states[None, top_x].reshape(-1, hidden_dim)
+            current_hidden_states = expert_layer(current_state) * routing_weights[top_x, idx, None]
+
+            # However `index_add_` only support torch tensors for indexing so we'll use
+            # the `top_x` tensor here.
+            final_hidden_states.index_add_(0, top_x, current_hidden_states.to(hidden_states.dtype))
+
+        # Add (optional) shared experts to the result
+        if self.shared_experts is not None:
```
Comment (ArthurZucker) [inline src/transformers/models/ernie4_5_moe/modular_ernie4_5_moe.py:189]: no we should not have optional codepaths

Comment (vasqu) [reply src/transformers/models/ernie4_5_moe/modular_ernie4_5_moe.py:189]: The problem is that the 300B does not have shared experts while the 21B does have them so it's necessary in this case.

File: src/transformers/models/ernie4_5_moe/modular_ernie4_5_moe.py (hunk 8 above)
Comment (vasqu) [reply src/transformers/models/ernie4_5_moe/modular_ernie4_5_moe.py:129]: See below

File: src/transformers/models/ernie4_5_moe/modular_ernie4_5_moe.py (hunk 7 above)
Comment (vasqu) [reply src/transformers/models/ernie4_5_moe/modular_ernie4_5_moe.py:125]: Hmm, I would have to convert weights then, no? I'd rather avoid weights conversion tbh since it would probably break support from other libs such as vllm.

Comment (vasqu) [reply src/transformers/models/ernie4_5_moe/modular_ernie4_5_moe.py:125]: Also since it follows mixtral and qwen3moe, it might be better to have this solution? https://github.com/huggingface/transformers/pull/39501#issuecomment-3089771718

Rather than forcing conversions here :eyes:

File: src/transformers/modeling_utils.py [hunk 10]
```diff
@@ -2977,6 +2977,17 @@ def _tie_or_clone_weights(self, output_embeddings, input_embeddings):
...
+        # Passing hooks over to the embeddings if needed
+        # (currently limited to tensor parallel hooks and flags only)
+        if hasattr(input_embeddings, "_is_hooked") and getattr(input_embeddings, "_hf_tp_plan", None):
+            output_embeddings._is_hooked = input_embeddings._is_hooked
+            output_embeddings._hf_tp_plan = input_embeddings._hf_tp_plan
+            output_embeddings._forward_hooks = input_embeddings._forward_hooks
+            output_embeddings._forward_pre_hooks = input_embeddings._forward_pre_hooks
+            output_embeddings.__repr__ = (
+                lambda: f"{output_embeddings.__repr__()}\nTP Plan: {output_embeddings._hf_tp_plan}"
+            )
```
Comment (ArthurZucker) [inline src/transformers/modeling_utils.py:2989]: okay! makes sense!

**Review by ArthurZucker — APPROVED** [review]

File: src/transformers/models/ernie4_5_moe/modular_ernie4_5_moe.py (hunk 7 above)
Comment (ArthurZucker) [reply src/transformers/models/ernie4_5_moe/modular_ernie4_5_moe.py:125]: no worries

File: src/transformers/models/ernie4_5_moe/modular_ernie4_5_moe.py (hunk 9 above)
Comment (ArthurZucker) [reply src/transformers/models/ernie4_5_moe/modular_ernie4_5_moe.py:189]: arf that's annoying
